File: src/models/File.js
import mongoose from "mongoose";

const fileSchema = new mongoose.Schema({
  sourceId: { type: mongoose.Schema.Types.ObjectId, ref: "Source", required: true, index: true },
  name: { type: String, required: true },
  path: { type: String, required: true },
  extension: { type: String, default: "" },
  mimeType: { type: String, default: "application/octet-stream" },
  category: {
    type: String,
    enum: ["text", "document", "image", "audio", "video", "other"],
    default: "other",
  },
  modelIngestible: { type: Boolean, default: false },
  size: { type: Number, default: 0 },
  createdAt: { type: Date },
  modifiedAt: { type: Date },
  metadata: {
    // text / document
    wordCount: Number,
    charCount: Number,
    // image
    width: Number,
    height: Number,
    ppi: Number,
    // audio
    duration: Number,
    bitrate: Number,
    // video
    fps: Number,
  },
}, { timestamps: false });

fileSchema.index({ sourceId: 1, path: 1 }, { unique: true });

export const File = mongoose.model("File", fileSchema);

File: backend/app/models/file.py
from datetime import datetime
from typing import Literal

import pymongo
from beanie import Document, PydanticObjectId
from pydantic import BaseModel, Field


FileCategory = Literal["text", "document", "image", "audio", "video", "other"]
RagIndexStatus = Literal["pending", "done", "failed", "skipped"]


class RagMetadata(BaseModel):
    indexStatus: RagIndexStatus = "pending"
    indexedChunks: int = 0
    lastIndexedAt: datetime | None = None
    extractor: str | None = None
    error: str | None = None


class FileMetadata(BaseModel):
    wordCount: int | None = None
    charCount: int | None = None
    width: int | None = None
    height: int | None = None
    ppi: float | None = None
    duration: int | None = None
    bitrate: int | None = None
    fps: float | None = None
    rag: RagMetadata | None = None


class FileDoc(Document):
    sourceId: PydanticObjectId
    name: str
    path: str
    extension: str = ""
    mimeType: str = "application/octet-stream"
    category: FileCategory = "other"
    size: int = 0
    createdAt: datetime | None = None
    modifiedAt: datetime | None = None
    metadata: FileMetadata = Field(default_factory=FileMetadata)

    class Settings:
        name = "files"
        indexes = [
            pymongo.IndexModel([("sourceId", pymongo.ASCENDING)]),
            pymongo.IndexModel(
                [("sourceId", pymongo.ASCENDING), ("path", pymongo.ASCENDING)],
                unique=True,
            ),
        ]
